content/lectures/lec02.md

```diff
@@ -6,6 +6,80 @@ assignments: ["hw0"]
+
+The layers also work as a cache. If we make a change to our container specification that only affects our source code, the underlying layers of the container, where we specify the OS and install dependencies like Apache *don't have to be rebuilt*. The container build process can simply rebuild the source code on top of the existing base image, which speeds up build time.
+
+Note that the successive images which we are layering on top only contain the differences. This saves us space initially (we only save the differences, so we don't store anything twice), but it also means that Docker images strictly increase in size as we continue to add layers on top. Be aware of this as you build your images: if you add something and then delete it in a later image, this still increases the size of the image because the differences are still stored.
+
+## Container Runtime
+
+An important distinction in what we've been talking about so far is what the difference is between a **container** and an **image**. An **image** is a blueprint for an instance of the code you want to run. A **container** is running instance of a Docker image. The Docker **runtime** is what takes an image, creates a container and runs it. 
+
+A Docker container is a running instance of a Docker image. Containers still employ a copy-on-write filesystem, layered off of the image they were spawned from. This means any new, updated or removed files during the lifetime of a container are *not* persisted after a container is stopped (Side note: the picture is a bit more complex in the original Docker implementation, but clarified with most other OCI-compatible runtimes). This means if we want to save any files that are created within a container while it runs, like log files, we have to do something other than save it to the filesystem.
+
+Docker runtime runs with a single host-wide Docker [daemon](https://en.wikipedia.org/wiki/Daemon_\(computing\)) (depicted in the earlier Docker visualization) which just sits in the background, manages your containers, and processes any requests which you might make with the Docker CLI to interact with the container. Docker's daemon is considered a single point of failure, and so other runtimes manage containers without a long-running, host-wide daemon process.
+
+## Publishing
+
+If we are sharing an image with someone, we don't want them to have to build the image from scratch every time. So, we need some way to publish a Docker image for someone else to use.
+
+The solution here is a registry where we can publish our images so that others can download them. Common examples include:
+- Docker Hub
+- Google Container Registry (GCR)
+- Elastic Container Registry (ECR)
+- GitHub Container Registry (unfortunately also abbreviated GCR)
+
+These are essentially the same thing as NPM or PIP, but for Docker images instead of code packages.
+
+### Tags
```

Comment:
We should explicitly mention that tags are just pointers to layers

Reply:
I didn't include this as a change because I don't really know how this works. What's the difference between being a pointer to an image and being a pointer to a layer?